### PR TITLE
[improvement](join) Share hash table in fragments for broadcast join

### PR DESCRIPTION
**Pull request by mrhhsg**
# Proposed changes

The hash table could be shared in join nodes of the same BE for broadcast join.
Hash join nodes with a shared hash table could significantly reduce memory and CPU usage.

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.31 seconds
 load time: 574 seconds
 storage size: 17154821251 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221102200011_clickbench_pr_38795.html

**Comment by yiguolei** [comment]
Great！！！

File: be/src/runtime/fragment_mgr.cpp [hunk 1]
```diff
@@ -738,6 +745,34 @@ void FragmentMgr::_set_scan_concurrency(const TExecPlanFragmentParams& params,
...
+    if (!params.__isset.fragment || !params.fragment.__isset.plan ||
+        params.fragment.plan.nodes.empty()) {
+        return;
+    }
+
+    for (auto& node : params.fragment.plan.nodes) {
+        if (node.node_type != TPlanNodeType::HASH_JOIN_NODE || !node.__isset.hash_join_node) {
+            continue;
+        }
+        if (node.runtime_filters.empty()) {
```
Comment (HappenLee) [inline be/src/runtime/fragment_mgr.cpp:760]: why the broadcast join node must have rf ？ rethink the logic？

File: be/src/vec/runtime/shared_hash_table_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    DCHECK(_hash_table_entries.find(my_node_id) == _hash_table_entries.cend());
+    _hash_table_entries.insert({my_node_id, std::move(entry)});
+    _cv.notify_all();
+}
+
+SharedHashTableEntry& SharedHashTableController::wait_for_hash_table(int my_node_id) {
+    std::unique_lock<std::mutex> lock(_mutex);
+    auto it = _hash_table_entries.find(my_node_id);
+    if (it == _hash_table_entries.cend()) {
+        _cv.wait(lock, [this, &it, my_node_id]() {
```
Comment (HappenLee) [inline be/src/vec/runtime/shared_hash_table_controller.cpp:46]: maybe `dead lock` when the node should be build hash table is canceled. it's dangerous

File: be/src/exprs/runtime_filter.cpp [hunk 3]
```diff
@@ -1108,7 +1141,8 @@ Status IRuntimeFilter::publish() {
         RETURN_IF_ERROR(
                 _state->runtime_filter_mgr()->get_consume_filter(_filter_id, &consumer_filter));
         // push down
-        std::swap(this->_wrapper, consumer_filter->_wrapper);
+        //        std::swap(this->_wrapper, consumer_filter->_wrapper);
```
Comment (HappenLee) [inline be/src/exprs/runtime_filter.cpp:1144]: delete the useless code

File: be/src/vec/runtime/shared_hash_table_controller.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+void SharedHashTableController::acquire_ref_count(RuntimeState* state, int my_node_id) {
+    std::unique_lock<std::mutex> lock(_mutex);
+    _ref_fragments[my_node_id].emplace_back(state->fragment_instance_id());
+}
+
+Status SharedHashTableController::release_ref_count(RuntimeState* state, int my_node_id) {
+    std::unique_lock<std::mutex> lock(_mutex);
+    RETURN_IF_CANCELLED(state);
```
Comment (HappenLee) [inline be/src/vec/runtime/shared_hash_table_controller.cpp:70]: if here return ？ still will be dead lock？

File: be/src/vec/runtime/shared_hash_table_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+void SharedHashTableController::_put_an_empty_entry_if_need(Status status, TUniqueId fragment_id,
+                                                            int node_id) {
+    auto builder_it = _builder_fragment_ids.find(node_id);
+    if (builder_it != _builder_fragment_ids.end()) {
+        if (builder_it->second == fragment_id) {
+            if (_hash_table_entries.find(builder_it->first) == _hash_table_entries.cend()) {
+                _hash_table_entries.insert(
+                        {builder_it->first, SharedHashTableEntry::empty_entry_with_status(status)});
+                LOG(INFO) << "here put an empty SharedHashTableEntry to avoid deadlocking";
```
Comment (HappenLee) [inline be/src/vec/runtime/shared_hash_table_controller.cpp:88]: seems be comment is better？

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**